### PR TITLE
add lint and test on pull_request

### LINKED ISSUE #1539: CMake Error at CMakeLists.txt:491 (get_flags): get_flags Function invoked with incorrect arguments for function named: get_flags
It seems that there is an issue when trying to build on linux, at least in our runner environment:

```
-- Could not find nvcc, please set CUDAToolkit_ROOT.
  CMake Warning at CMakeLists.txt:318 (message):
    cuBLAS not found
  
  
  -- CUDA host compiler is GNU 
  CMake Error at CMakeLists.txt:491 (get_flags):
    get_flags Function invoked with incorrect arguments for function named:
    get_flags
  
  
  -- CMAKE_SYSTEM_PROCESSOR: x86_64
  -- x86 detected
  -- Configuring incomplete, errors occurred!
  llm/llama.cpp/generate_linux.go:24: running "cmake": exit status 1
```

See the following issue for more details:

- https://github.com/Homebrew/homebrew-core/pull/157426

If this is an issue in our environment, that's fair enough. I mostly wanted to make sure there was an issue captured here for the error message to assist others in future.

### PR DESCRIPTION
**Pull request by mxyng**
fixes a bug with generate where `get_flags` errors on ubuntu (and likely windows) when building cuda on a cuda-less system

fixes a bug in windows where `/api/list` does not return models correctly

both lint and test requires go generate results so do it once then propagate the artifacts to the rest of the pipeline

this enables linting with golangci-lint but doesn't go overboard with linters. it enables the default linters and only a few extra linters to catch the most egregious bugs

linting can run locally with [golangci-lint](https://golangci-lint.run/usage/install/)

```
go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.55.2
golangci-lint run -v
```

```
docker run --rm -v $(pwd):/app -w /app golangci/golangci-lint:v1.55.2 golangci-lint run -v
```

resolves #1539

### CODE REVIEW

File: .github/workflows/test.yaml
Comment (mxyng) [inline .github/workflows/test.yaml:32]: `cache: false` here because golangci-lint is finicky

File: server/modelpath.go
Comment (mxyng) [inline server/modelpath.go:49]: This is a URL path, not a file path so `os.PathSeparator` is inappropriate

File: readline/history.go
Comment (mxyng) [inline readline/history.go:135]: I'm not sure why this is 666 @pdevine

File: server/routes.go
Comment (mxyng) [inline server/routes.go:713]: This normalizes the modelpath for windows

File: readline/history.go
Comment (pdevine) [reply readline/history.go:135]: Not sure... it's set to 600 above here. Probably best to change it.

File: .github/workflows/test.yaml
Comment (BruceMacD) [inline .github/workflows/test.yaml:38]: is the uploading on downloading of artifacts much overhead? Im wondering if it would be more efficient to just do the `generate` and `test` as one job

File: .golangci.yaml
Comment (BruceMacD) [inline .golangci.yaml:12]: Why are these functions excluded? Might be worth a comment here

File: llm/llama.cpp/generate_darwin_amd64.go
Comment (BruceMacD) [inline llm/llama.cpp/generate_darwin_amd64.go:16]: these warnings break the CI `generate` step?

**Review by BruceMacD — COMMENTED** [review]
feels great to see some of these little bugs cleaned up

File: .github/workflows/test.yaml
Comment (mxyng) [reply .github/workflows/test.yaml:38]: It's not and definitely faster than running generate in each job. v4 is also much faster according to their readme

File: llm/llama.cpp/generate_darwin_amd64.go
Comment (mxyng) [reply llm/llama.cpp/generate_darwin_amd64.go:16]: yes, in particular the cmake function `get_flags` on linux without cuda

**Review by BruceMacD — APPROVED** [review]
Jeff and Patrick will probably want to take a look too, but this looks good to me.

File: .github/workflows/test.yaml
Comment (dhiltgen) [inline .github/workflows/test.yaml:23]: `llama.cpp/gguf/build/lib/*` is where the dynamic loaded libraries go.
`llama.cpp/gguf/ggml-metal.metal` is where we'll pull the metal file for darwin

The "built in" version for linux links against the archives from the CPU build variant located here:
`llm/llama.cpp/gguf/build/cpu/examples/server/libext_server.a`
`llm/llama.cpp/gguf/build/cpu/common/libcommon.a`
`llm/llama.cpp/gguf/build/cpu/libllama.a`
`llm/llama.cpp/gguf/build/cpu/libggml_static.a`

The built-in darwin is linked against these
`llm/llama.cpp/gguf/build/metal/common/libcommon.a`
`llm/llama.cpp/gguf/build/metal/examples/server/libext_server.a`
`llm/llama.cpp/gguf/build/metal/libllama.a`
`llm/llama.cpp/gguf/build/metal/libggml_static.a`


I can see this is a bit messy, so we may want to follow up with some post-processing step(s) in the generate script to aggregate these into a more consistent/clean location.

Comment (dhiltgen) [reply .github/workflows/test.yaml:23]: Additional breadcrumbs:
* https://github.com/jmorganca/ollama/blob/main/llm/shim_darwin.go#L12
* https://github.com/jmorganca/ollama/blob/main/llm/shim_ext_server.go#L28
* https://github.com/jmorganca/ollama/blob/main/llm/ext_server.go#L13-L16
* https://github.com/jmorganca/ollama/blob/main/llm/ext_server.go#L20-L23

File: .github/workflows/test.yaml
Comment (chenrui333) [inline .github/workflows/test.yaml:16]: maybe good to use go1.21?

Comment (mxyng) [reply .github/workflows/test.yaml:16]: Ollama currently targets go1.20 for compatibility so that's what's being tested here

File: .github/workflows/test.yaml
Comment (BruceMacD) [inline .github/workflows/test.yaml:24]: what does this do?

File: .github/workflows/test.yaml
Comment (BruceMacD) [inline .github/workflows/test.yaml:4]: Should we set this to be only runnable by maintainers somehow? I am wondering if the `go generate ./...` step will use a lot of our runner minutes.

File: .github/workflows/test.yaml
Comment (BruceMacD) [inline .github/workflows/test.yaml:77]: Do we need to generate and build? I'm wondering if `go test` and the linter would be sufficient.

File: .github/workflows/test.yaml
Comment (jmorganca) [reply .github/workflows/test.yaml:4]: If it's fast enough <5min, it's okay if it's ok even on all commits

File: .github/workflows/test.yaml
Comment (mxyng) [reply .github/workflows/test.yaml:24]: This sets the environment variables in Windows such that additional Visual Studio tools like `dumpbin` used in gen_window.ps1 are in the path. Usually a shell created in Visual Studio will automatically export these variables but since this isn't Visual Studio, we need to do it manually

File: .github/workflows/test.yaml
Comment (mxyng) [reply .github/workflows/test.yaml:4]: There's additional controls at the repo level where it will require maintainer approval before running. I'm not sure if that's enabled but it's another knob to turn

File: .github/workflows/test.yaml
Comment (mxyng) [reply .github/workflows/test.yaml:77]: We definitely need to generate because lint and test requires the generate artifacts; `go:embed` errors if those files aren't available. Additionally, it's useful to catch potential issues with the generate process

build is trivial once the artifacts are available

**Review by BruceMacD — APPROVED** [review]

**Pull request merged**
